### DISCUSSION TITLE
Too frequent DNS requests

### DISCUSSION THREAD

**Discussion by supply9243** [rustdesk/rustdesk#5896] [General]
### Bug Description

When the server domain name does not have AAAA resolution, the Windows client will request resolution once every 15 seconds on average, reaching 5,500 requests in 24 hours. No other program requests parsing so frequently.

当服务器域名没有AAAA解析时，Windows客户端会以平均15秒一次请求解析，24小时内达到了5500请求数.
没有哪个程序会如此频繁的请求解析，

### How to Reproduce

Use AdGuard Home to view logs while the software is running.

软件运行时使用 AdGuard Home 查看日志。

### Expected Behavior

When one of IPv6/IPv4 can be connected, there is no need to request resolution of another protocol, or reduce its frequency to follow domain name TTL rules.

当IPv6/IPv4有一个可以连接时，不必再请求另一个协议的解析，或者降低其频率遵循域名TTL规则。

### Operating system(s) on local side and remote side

Windows10

### RustDesk Version(s) on local side and remote side

1.2.2

### Screenshots

![2023-10-03_14-34-37](https://github.com/rustdesk/rustdesk/assets/50667472/72645c1d-983b-4c76-abba-9bbae7f796f0)
![2023-10-03_14-34-59](https://github.com/rustdesk/rustdesk/assets/50667472/2f552fce-2c17-45f2-b3a2-9a80dc558126)

**Comment by supply9243** [comment]
This is just a request from one device.
这只是一台设备的请求。

**Comment by ianalexis** [comment]
Same reported here. #8984 
I dont think this many requests are acceptable.

**Comment by KayLeung** [comment]
这问题被拒绝修正，但没有正面解析原因？ （查看了几个 ISSUES）
PRO 版好像没有这问题
也就是说 PRO 版接收请求后就终止？
不清楚 API 用途，即使 PRO 版是不是也在需求时才发出？
假设维持状态还是什么，也太多了吧...

smiliar problem: #13938

我已经停用，先不说隐私问题，因为这就像一种 DDoS 攻击...

希望是我理解错误...
-

This issue was closed as "wontfix," but there hasn't been a direct explanation as to why. (I’ve looked through several related issues.)

It seems the PRO version doesn't have this problem. Does that mean the PRO version STOPPED after receiving a response?

Assumed PRO version does it also, shouldn't it only triggered on demand? Even if it's for maintaining state or something similar, the frequency is simply excessive...

Similar problem: #13938

I have already stopped to use rustdesk Privacy concerns aside, the behavior is essentially like a DDoS attack.

**Comment by XXXBold** [comment]
I also experience the same issue, I used AI (Claude Opus 4.8) to lookup the code and find what might cause this.
I'm ready to investigate further here if there's green light from the maintainer that they will actually accept a fix for this (not so sure because other issues were just closed).

AI Findings:

> **Disclosure:** the source-code analysis below was produced with the help of an AI assistant (Claude). I've read through it against the current `master` and it matches what I see, but treat the specifics as "please verify against the tree you're building" rather than gospel — symbol names are stable, exact line numbers may drift between versions.

I went through the source to find what's actually generating these lookups, since this has now been reported a few times (#5893 → this discussion, #8984, and more recently #13938 with the 443 / 21114 traffic) without a root cause being posted. The DNS flood here and the "pings every 15s on 443/21114" in #13938 are the **same** thing.

## Root cause

It's the HTTP heartbeat in `src/hbbs_http/sync.rs` (`start_hbbs_sync_async`), not the rendezvous/relay protocol itself.

The task's interval ticks every 3s (`TIME_CONN`), but when idle it's gated to one heartbeat per 15s (`TIME_HEARTBEAT`):

```rust
const TIME_HEARTBEAT: Duration = Duration::from_secs(15);
// ...
if conns.is_empty() && last_sent.map(|x| x.elapsed() < TIME_HEARTBEAT).unwrap_or(false) {
    continue;
}
last_sent = Some(Instant::now());
// ... crate::post_request(url, ...).await   // POST {api-server}/api/heartbeat
```

So idle → one POST to `{api-server}/api/heartbeat` every 15s. (During an active session the gate is lifted and it's effectively every 3s.)

**Why only on self-hosted / custom servers** — `heartbeat_url()` bails out for the public server:

```rust
let url = crate::common::get_api_server(
    Config::get_option("api-server"),
    Config::get_option("custom-rendezvous-server"),
);
if url.is_empty() || crate::is_public(&url) {   // is_public matches rustdesk.com only
    return "".to_owned();
}
```

And when `api-server` isn't set explicitly, `get_api_server_()` derives it from `custom-rendezvous-server`:

```rust
let s = crate::increase_port(&s0, -2);          // RENDEZVOUS_PORT - 2 = 21114
// -> "http://{host}:21114"   (or https:443 if you front it with a proxy)
```

That's exactly the **port 443 + intermittent 21114** traffic reported in #13938.

**Why it floods DNS (the symptom in this thread)** — the POST path is `post_request` → `post_request_http` → `post_request_` (`src/common.rs`), and on **every single call** it builds a brand-new HTTP client:

```rust
let mut req = create_http_client_async(/* ... */).post(url);
```

`create_http_client_async` (`src/hbbs_http/http_client.rs`) does `AsyncClient::builder()...build()` each time — a fresh `reqwest::Client` with its own connection pool that's dropped when the function returns. There's no shared/cached client, no `OnceCell`/`lazy_static` around it, and no application-level resolver cache. So every 15s opens a cold connection, which forces a fresh `getaddrinfo` → an **A and an AAAA query** each cycle. That's the ~2 lookups / 15s (≈5,500/day) seen here, and it's most wasteful exactly as the OP described: when the domain has no AAAA record, the AAAA half never resolves to anything useful but still gets re-queried every cycle.

For completeness: the rendezvous mediator (`rendezvous_mediator.rs`) has its **own** 15s heartbeat (`REG_INTERVAL = 15_000`), but that one reuses the already-resolved UDP/TCP socket and only re-resolves on failure (UDP path even throttles that to `DNS_INTERVAL = 60s`). So it is **not** the source of these lookups — the HTTP heartbeat building a throwaway client is.

## Suggested fixes (either one largely solves it)

1. **Reuse the client.** Build one `reqwest::Client` once (behind `OnceCell`/`lazy_static`, keyed by TLS settings if needed) and reuse it for the heartbeat. reqwest keeps the connection in its pool, so an idle client re-resolves nothing between 15s heartbeats. This is the cleanest fix and also reduces TCP/TLS churn, not just DNS.
2. **Stop re-querying both families / honor TTL**, per the original request in this thread — if A connects, don't keep firing AAAA every cycle.

(1) is the higher-leverage change because it eliminates the per-request connect entirely while idle. Happy to test a patch against current `master` if a maintainer wants to take a look — this is reproducible 1:1 with a self-hosted OSS server and any local DNS resolver (AdGuard Home / Pi-hole) showing the per-client query log.
